app/plugin_install.go

```diff
@@ -103,53 +118,52 @@ func (a *App) installPluginLocally(pluginFile io.ReadSeeker, replace bool) (*mod
+		return nil, model.NewAppError("installPluginLocally", "app.plugin.invalid_id.app_error", map[string]interface{}{"Min": plugin.MinIdLength, "Max": plugin.MaxIdLength, "Regex": plugin.ValidIdRegex}, "", http.StatusBadRequest)
 	}
 
-	// Stash the previous state of the plugin, if available
-	stashed := a.Config().PluginSettings.PluginStates[manifest.Id]
-
 	bundles, err := pluginsEnvironment.Available()
 	if err != nil {
-		return nil, model.NewAppError("installPlugin", "app.plugin.install.app_error", nil, err.Error(), http.StatusInternalServerError)
+		return nil, model.NewAppError("installPluginLocally", "app.plugin.install.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
 	// Check that there is no plugin with the same ID
 	for _, bundle := range bundles {
 		if bundle.Manifest != nil && bundle.Manifest.Id == manifest.Id {
 			if !replace {
-				return nil, model.NewAppError("installPlugin", "app.plugin.install_id.app_error", nil, "", http.StatusBadRequest)
+				return nil, model.NewAppError("installPluginLocally", "app.plugin.install_id.app_error", nil, "", http.StatusBadRequest)
 			}
 
 			if err := a.removePluginLocally(manifest.Id); err != nil {
-				return nil, model.NewAppError("installPlugin", "app.plugin.install_id_failed_remove.app_error", nil, "", http.StatusBadRequest)
+				return nil, model.NewAppError("installPluginLocally", "app.plugin.install_id_failed_remove.app_error", nil, "", http.StatusBadRequest)
 			}
 		}
 	}
 
 	pluginPath := filepath.Join(*a.Config().PluginSettings.Directory, manifest.Id)
 	err = utils.CopyDir(tmpPluginDir, pluginPath)
 	if err != nil {
-		return nil, model.NewAppError("installPlugin", "app.plugin.mvdir.app_error", nil, err.Error(), http.StatusInternalServerError)
+		return nil, model.NewAppError("installPluginLocally", "app.plugin.mvdir.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
 	// Flag plugin locally as managed by the filestore.
 	f, err := os.Create(filepath.Join(pluginPath, managedPluginFileName))
 	if err != nil {
-		return nil, model.NewAppError("uploadPlugin", "app.plugin.flag_managed.app_error", nil, err.Error(), http.StatusInternalServerError)
+		return nil, model.NewAppError("installPluginLocally", "app.plugin.flag_managed.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 	f.Close()
 
-	if stashed != nil && stashed.Enable {
-		a.EnablePlugin(manifest.Id)
-	}
-
-	if err := a.notifyPluginStatusesChanged(); err != nil {
-		mlog.Error("failed to notify plugin status changed", mlog.Err(err))
+	// Activate plugin if it was previously activated.
+	pluginState := a.Config().PluginSettings.PluginStates[manifest.Id]
+	if pluginState != nil && pluginState.Enable {
+		updatedManifest, _, err := pluginsEnvironment.Activate(manifest.Id)
+		if err != nil {
+			return nil, model.NewAppError("installPluginLocally", "app.plugin.restart.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+		manifest = updatedManifest
```

Comment:
To confirm, this is new in the diff -- returning the updated manifest -- what do we expect to have changed to return, exactly?

Reply:
Ahh this was another gotcha i ran into. When activating a plugin, there is code that updates the manifest in memory that has the updated webapp bundle hash. Done here: https://github.com/mattermost/mattermost-server/blob/c1c2da804f088ba754b3fdce24caca4a697d4a18/plugin/environment.go#L246

Whereas fetching manifest through `pluginsEnvironment.Available()` even after a plugin is enabled, does not have the up to date webapp bundle hash as that is only in memory and not persisted to disk, which is why i have to do this, rather than just getting the upto date manifest